protocol/contracts/lpp/src/state/total.rs

```diff
@@ -102,19 +110,27 @@ where
         loan_principal_payment: Coin<Lpn>,
         loan_interest_rate: Percent,
     ) -> &Self {
-        self.total_interest_due = self.total_interest_due_by_now(&ctime) - loan_interest_payment;
+        // The interest payment calculation of loans is the source of truth.
+        // Therefore, it is possible for the rounded-down total interest due from `total_interest_due_by_now`
+        // to become less than the sum of loans' interests. Taking 0 when subtracting a loan's interest from the total is a safe solution.
 
-        self.annual_interest_rate = if self.total_principal_due == loan_principal_payment {
-            zero_interest_rate()
-        } else {
+        self.total_interest_due = self
+            .total_interest_due_by_now(&ctime)
+            .saturating_sub(loan_interest_payment);
+
+        self.annual_interest_rate = if loan_principal_payment < self.total_principal_due {
             Rational::new(
                 Fraction::<Coin<Lpn>>::of(&self.annual_interest_rate, self.total_principal_due)
                     - loan_interest_rate.of(loan_principal_payment),
                 self.total_principal_due - loan_principal_payment,
             )
+        } else {
+            zero_interest_rate()
         };
 
-        self.total_principal_due -= loan_principal_payment;
+        self.total_principal_due = self
+            .total_principal_due
+            .saturating_sub(loan_principal_payment);
```

Comment:
here we do not have any rounding that may cause an overflow

It is a bug if we get an overflow. Therefore we should panic, not error, if that happens.
Hint: Use `checked_sub` and `expect` to panic with a custom message explaining the case